File: csharp/AvoidHadoukenIfs.cs
// Avoid Hadouken IFs
// Avoid nested IFs (Hadouken IFs)
// Solution: Early Returns, Switch-Cases

using System;

public class AvoidHadoukenIfs
{
    public class Order
    {
        private bool isValid = true;
        private bool isInStock = true;
        private bool paymentSuccess = true;

        public bool Valid()
        {
            return isValid;
        }

        public bool InStock()
        {
            return isInStock;
        }

        public bool PaymentSuccessful()
        {
            return paymentSuccess;
        }
    }

    // Before
    public static void ProcessOrderBefore(Order order)
    {
        if (order.Valid())
        {
            if (order.InStock())
            {
                if (order.PaymentSuccessful())
                {
                    Console.WriteLine("Order processed successfully!");
                }
                else
                {
                    Console.WriteLine("Payment failed.");
                }
            }
            else
            {
                Console.WriteLine("Item is out of stock.");
            }
        }
        else
        {
            Console.WriteLine("Order is invalid.");
        }
    }

    // After
    public static void ProcessOrder(Order order)
    {
        if (!order.Valid())
        {
            Console.WriteLine("Order is invalid.");
            return;
        }
        
        if (!order.InStock())
        {
            Console.WriteLine("Item is out of stock.");
            return;
        }
        
        if (!order.PaymentSuccessful())
        {
            Console.WriteLine("Payment failed.");
            return;
        }

        Console.WriteLine("Order processed successfully!");
    }

    public static void Main(string[] args)
    {
        var order = new Order();
        
        ProcessOrderBefore(order);
        Console.WriteLine();
        ProcessOrder(order);
    }
}

File: python/avoid_hadouken_ifs.py
# Avoid Hadouken IFs
# Avoid nested IFs (Hadouken IFs)
# Solution: Early Returns, Switch-Cases

# Before
class Order:
    def __init__(self):
        self.is_valid = True
        self.is_in_stock = True
        self.payment_success = True

    def valid(self):
        return self.is_valid

    def in_stock(self):
        return self.is_in_stock

    def payment_successful(self):
        return self.payment_success

def process_order(order):
    if order.valid():
        if order.in_stock():
            if order.payment_successful():
                print("Order processed successfully!")
            else:
                print("Payment failed.")
        else:
            print("Item is out of stock.")
    else:
        print("Order is invalid.")

order = Order()
process_order(order)


# After
class Order:
    def __init__(self):
        self.is_valid = True
        self.is_in_stock = True
        self.payment_success = True

    def valid(self):
        return self.is_valid

    def in_stock(self):
        return self.is_in_stock

    def payment_successful(self):
        return self.payment_success

def process_order(order):
    if not order.valid():
        print("Order is invalid.")
        return

    if not order.in_stock():
        print("Item is out of stock.")
        return

    if not order.payment_successful():
        print("Payment failed.")
        return

    print("Order processed successfully!")

order = Order()
process_order(order)
